test/shared/kaocha/integration_helpers.clj

```diff
@@ -126,6 +129,123 @@
+    (binding [*process* p]
+      (try
+        (let [input (.getOutputStream p)
+              output (.getInputStream p)]
+          (with-open [w (io/writer input)]
+            (with-open [r (io/reader output)]
+              (future
+                ;; unblock read-line calls after 30 seconds and abandon test
+                (Thread/sleep timeout-ms)
+                @kill)
+              (binding [*in* r
+                        *out* w]
+                (f)))))
+        (assert (.waitFor p 10 (java.util.concurrent.TimeUnit/SECONDS))
+                "Process failed to stop!")
+        (.exitValue p)
+        (finally
+          (is (not (realized? kill)) (format "Process was killed after %sms timeout!!" timeout-ms))
+          (try (.exitValue p)
+               (catch IllegalThreadStateException _
+                 (is nil "Process was killed after executing entire test suite!!")
+                 @kill)))))))
+
+(defmacro interactive-process
+  "Simulate a test run of a bin/kaocha integration test
+  using (read-line) and (println).
+
+  m is the output from `test-dir-setup`
+  args is a collection of arguments after ./bin/kaocha.
+  eg., [\"--watch\" \"--focus\" \"foo\"]
+  
+  Kills the process after 30 seconds if it has not already
+  been exited. Ensures the process is killed immediately after
+  body is executed. Returns the exit code.
+  
+  Executes body in the following environment:
+  *process* is the running Process executing `./bin/kaocha ~@args`
+  *in* is bound to the output stream of this Process
+  - eg., use (read-line) to read the output of ./bin/kaocha
+  *out* is bound to the input stream of this Process
+  - eg., use (println) to send input to ./bin/kaocha
+  
+  Tips:
+  - avoid using `testing` or `is` in ways that can swallow exceptions
+  in body. Try and make the first exception that's thrown contain
+  all the debugging information you need using the helpers in the
+  rest of this file."
+  [m args & body]
+  `(interactive-process* ~m ~args #(do ~@body)))
+
+(comment
+  (zero?
+    (interactive-process {:dir "." :runner (io/file "echo")} ["hello"]
+                         (let [l1 (read-line)
+                               _ (assert (= "hello" l1) (pr-str l1))])))
+  )
+
+(defn read-string-line
+  "Read a line from the current integration test and throw if the process has died."
+  []
+  (or (read-line)
+      (throw (ex-info "Process killed!!" {}))))
+
+(defn expect-lines
+  "Assert that lines, a vector of strings, matches the next lines from the integration process.
+
+  If not, slurps the rest of the output for debugging purposes and throws an exception."
+  [lines]
+  (mapv (fn [l]
+          (let [s (read-string-line)]
+            (or (is (= l s))
+                (throw (ex-info (format "Failed to match %s\nEntire expected: %s\nRest of stream:\n%s"
+                                        (pr-str l) lines (str/split-lines (slurp *in*)))
+                                {})))))
+        lines))
```

Comment:
Could `match?` (from matchers/combinators) be used instead of this function? `kaocha.watch-test` already uses this in `reload-config-test`.

Reply:
I've never used it, but I changed it for you.